torch/utils/tensorboard/_pytorch_graph.py

```diff
@@ -0,0 +1,267 @@
+        self.tensor_size = tensor_size
+        self.kind = 'Parameter'
+        if input_or_output:
+            self.input_or_output = input_or_output
+            self.kind = 'IO Node'
+
+
+class NodePyOP(NodePy):
+    def __init__(self, NodeCpp):
+        super(NodePyOP, self).__init__(NodeCpp, methods_OP)
+        self.attributes = str({k: NodeCpp[k] for k in NodeCpp.attributeNames()}).replace("'", ' ')
+        self.kind = NodeCpp.kind()
+
+
+class GraphPy(object):
+    def __init__(self):
+        self.nodes_op = []
+        self.nodes_io = OrderedDict()
+        self.uniqueNameToScopedName = {}
+        self.shallowestScopeName = 'default'
+        self.scope_name_appeared = []
+
+    def append(self, x):
+        if type(x) == NodePyIO:
+            self.nodes_io[x.uniqueName] = x
+        if type(x) == NodePyOP:
+            self.nodes_op.append(x)
+            for node_output, outputSize in zip(x.outputs, x.outputstensor_size):
+                self.scope_name_appeared.append(x.scopeName)
+                self.nodes_io[node_output] = NodeBase(node_output,
+                                                       x.inputs,
+                                                       x.scopeName,
+                                                       outputSize,
+                                                       op_type=x.kind,
+                                                       attributes=x.attributes)
+
+    def printall(self):
+        print('all nodes')
+        for node in self.nodes_op:
+            print(node)
+        for key in self.nodes_io:
+            print(self.nodes_io[key])
+
+    def find_common_root(self):
+        for fullscope in self.scope_name_appeared:
+            if fullscope:
+                self.shallowestScopeName = fullscope.split('/')[0]
+
+    def populate_namespace_from_OP_to_IO(self):
+        for node in self.nodes_op:
+            for input_node_id in node.inputs:
+                self.uniqueNameToScopedName[input_node_id] = node.scopeName + '/' + input_node_id
+
+        for key, node in self.nodes_io.items():
+            if type(node) == NodeBase:
+                self.uniqueNameToScopedName[key] = node.scope + '/' + node.uniqueName
+            if hasattr(node, 'input_or_output'):
+                self.uniqueNameToScopedName[key] = node.input_or_output + '/' + node.uniqueName
+            if hasattr(node, 'scope'):
+                if node.scope == '' and self.shallowestScopeName:
+                    self.uniqueNameToScopedName[node.uniqueName] = self.shallowestScopeName + '/' + node.uniqueName
+
+        # replace name
+        for key, node in self.nodes_io.items():
+            self.nodes_io[key].inputs = [self.uniqueNameToScopedName[node_input_id] for node_input_id in node.inputs]
+            if node.uniqueName in self.uniqueNameToScopedName:
+                self.nodes_io[key].uniqueName = self.uniqueNameToScopedName[node.uniqueName]
+
+    def to_proto(self):
```

Comment:
I was surprised to see `nodes_op` was not used at all in this method.

Reply:
...is this what `populate_namespace_from_OP_to_IO` does?

Reply:
@lanpa can you comment here? Thanks.

Reply:
The input node[s] used by each node `n` in `self.nodes_OP` is assigned the same scope of `n`. This is for better visualization. 
```python
for node in self.nodes_OP:
  for input_node_id in node.inputs:
    self.uniqueNameToScopedName[input_node_id] = node.scopeName + '/' + input_node_id
```

Reply:
Added task. Thanks.